avm2,audio: Implement `SampleDataEvent`

torokati44 opened the pull request:
Fixes: #14156
Also fixes z0r.de loops: 5885, 7024 (issue #7879), 7312
Probably fixes (to be tested): #9171, #13655

It's quite rough (full of `unwrap()`s for example, and overall unfinished) but it "just started working".

Mostly pushing to show that it's in progress - any ideas on how to improve the architecture are welcome.

An unexpected (to me at least) problem is that sound synthesis in AS3 needs to do quite a lot of math really fast, so on my machine, even in release mode, the content linked above could barely keep up with real time sound generation, or not at all. Profiling and optimization is yet to be done (probably related: #7920).

SN902 commented:
Most NSFW video games depend on this commit, here still no any SFX, but now they panics after the end of the first loop (need to wait for the issue reproduction)

[Twi_Armor.zip](https://github.com/ruffle-rs/ruffle/files/13510058/Twi_Armor.zip)

`thread 'main' panicked at 'not too many sounds: TooManySounds', core\src\avm2\globals\flash\media\sound.rs:179:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
AVM2 stack trace:
        at flash.media::Sound/flash::media::Sound::play()
        at mp3pl::MP3Player/play()
        at appleblitz_fla::MainTimeline/playsounds()
        at appleblitz_fla::MainTimeline/loopend()
        at appleblitz_fla::MainTimeline/frame1()
error: process didn't exit successfully: `target\release\ruffle_desktop.exe` (exit code: 0xc0000409, STATUS_STACK_BUFFER_OVERRUN)`

Lord-McSweeney commented on core/src/backend/audio/mixer.rs:
```diff
@@ -881,6 +899,68 @@ impl Stream for EventSoundStream {
+struct GeneratedSoundStream {
+    position: u32,
+    /// Contents of the shared `next_samples` are periodically moved into
+    /// to to reduce lock contention while yielding audio frames.
```
typo: "to to reduce"

torokati44 replied:
Thank you, fixed!

torokati44 commented:
TBD:
 - verify that the architecture is correct - the `load_called` method, how `AudioManager` and `AudioBackend` and `Mixer` keep track of what, the new stream types
 - test all the games/animations mentioned above, and actually make them work properly
 - some kind of regression test
 - see if any more events need to be dispatched

Lord-McSweeney commented on core/src/backend/audio.rs:
```diff
@@ -425,22 +441,72 @@ impl<'gc> AudioManager<'gc> {
+
+                ba.set_position(0);
+                for _ in 0..ba.len() / 4 {
+                    ns.push_back(ba.read_float().expect("float can be read"));
```
Shouldn't failing to read a float result in the read loop terminating early?

torokati44 replied:
that is to be tested :|

torokati44 replied:
AFAIK `read_float` can only fail if there are not enough bytes (any four bytes can be reinterpreted as an f32), and the `ba.len() / 4` in the loop ensures this never happens.

torokati44 replied:
added a comment about this nevertheless (just not pushed yet)

torokati44 commented:
@copilot try to create a new PR inspired by this, implemented on top of current `master`. like a git rebase, but it will be complicated...

torokati44 committed:
avm2: Add SampleDataEvent constructor and native data access

Add a native `sample_data_event` constructor on `EventObject`,
register the `sampleData` common string, and mark
`SampleDataEvent._data` as `[Ruffle(NativeAccessible)]` so the
audio backend can read back the samples written by ActionScript.

torokati44 committed:
avm2: Add Empty/Loading/Generated SoundData states

Rename `SoundData::NotLoaded` to `Loading`, and add `Empty`
(initial state) and `Generated` (when `play()` is called before
`load()`) states. Wire `Sound.play`/`Sound.load` to register a
generated sound stream when playback starts on an unloaded sound.

torokati44 committed:
audio: Add support for generated sound streams

Support sounds synthesized in ActionScript: each frame the
mixer requests more samples, which are produced by dispatching
a `SampleDataEvent` to the associated AS3 `Sound` object and
reading back its `data` ByteArray.
Re-export `SoundObject` from the `avm2` module so it can be
referenced by the audio backend when handling generated sounds.

torokati44 committed:
tests: Add test for SampleDataEvent generated sounds

Plays an unloaded `Sound` with a `SAMPLE_DATA` listener, synthesizing
a 0.5-amplitude 440 Hz sine for ~0.5 seconds and then stopping.
Audio volume assertions verify the mixed output peaks at ~0.5 while
the tone plays, and is completely silent once the generator stops
supplying samples - exercising the full generated-sound
`SampleDataEvent` path end-to-end.

Lord-McSweeney commented on core/src/avm2/globals/flash/media/sound.rs:
```diff
@@ -174,6 +174,33 @@ pub fn play<'gc>(
+                .context
+                .audio_manager
+                .start_generated_sound(activation.context.audio, sound_object)
+                .expect("not too many sounds");
```
Can't this `expect` be hit if too many sounds are generated?

torokati44 replied:
Yep, and we should return `null` to AS instead of panicking. Good call!

Lord-McSweeney commented on core/src/avm2/object/sound_object.rs:
```diff
@@ -161,6 +169,38 @@ impl<'gc> SoundObject<'gc> {
+                };
+            }
+            SoundData::Loading { .. } => {
+                panic!("Tried to load sound that is already Loading");
```
Is there really no way for these panics to be hit? If not, could you add explanatory comments?

torokati44 replied:
Ah, there are a couple of these cases that should be AS exceptions instead, will look into it!

torokati44 committed:
tests: Add a test with too many generated (SampleDataEvent) sounds

torokati44 committed:
tests: Add a test for calling Sound.play twice and then Sound.load

torokati44 commented on core/src/avm2/object/sound_object.rs:
```diff
@@ -120,14 +122,15 @@ pub enum SoundLoadingState {
     New,
     Loading,
     Loaded,
+    Generated,
 }
```
Are `SoundData` and `SoundLoadingState` supposed to mirror each other like this? @kjarosh

I'm asking you because you added the latter in https://github.com/ruffle-rs/ruffle/pull/19586 .

~~(And now that I'm reading it, maybe `SoundData::NotLoaded` and `SoundData::Loading` should be collapsed into one - it's probably split as the result of another glitchy rebase from eons ago...)~~ EDIT: lol I read the diff wrong when writing this, there's no split like this

kjarosh replied:
Well, it depends on whether sound behaves the same way in AVM1 and AVM2. If it does and we have a lot of evidence of that, perhaps it's a good idea to unify the structures between AVM1 and AVM2 to some extent?

torokati44 replied:
It's strange though because SoundData is used from both avm1 and avm2, while SoundLoadingState is only used for/by avm2.
